src/util/time.cpp

```diff
@@ -9,39 +9,135 @@
 #endif
 
 #include "util/time.h"
+#include "sync.h"
 
 #include <chrono>
 #include <boost/date_time/posix_time/posix_time.hpp>
 #include <boost/thread.hpp>
 
-using namespace std;
+RecursiveMutex cs_clock;
+static CClock* zcashdClock = SystemClock::Instance();
 
-static int64_t nMockTime = 0; //!< For unit testing
-
-int64_t GetTime()
-{
-    if (nMockTime) return nMockTime;
-
-    return time(NULL);
+void SystemClock::SetGlobal() {
+    LOCK(cs_clock);
+    zcashdClock = SystemClock::Instance();
 }
 
-void SetMockTime(int64_t nMockTimeIn)
-{
-    nMockTime = nMockTimeIn;
+int64_t SystemClock::GetTime() const {
+    return std::chrono::duration_cast<std::chrono::seconds>(
+            std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
-int64_t GetTimeMillis()
-{
+int64_t SystemClock::GetTimeMillis() const {
     return std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
-int64_t GetTimeMicros()
-{
+int64_t SystemClock::GetTimeMicros() const {
     return std::chrono::duration_cast<std::chrono::microseconds>(
             std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
+void FixedClock::SetGlobal(std::chrono::seconds fixedSeconds) {
+    LOCK(cs_clock);
+    FixedClock::Instance()->Set(fixedSeconds);
+    zcashdClock = FixedClock::Instance();
+}
+
+int64_t FixedClock::GetTime() const {
+    return fixedSeconds.count();
+}
+
+int64_t FixedClock::GetTimeMillis() const {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(fixedSeconds).count();
+}
+
+int64_t FixedClock::GetTimeMicros() const {
+    return std::chrono::duration_cast<std::chrono::microseconds>(fixedSeconds).count();
+}
+
+void OffsetClock::SetGlobal(std::chrono::seconds offsetSeconds) {
+    LOCK(cs_clock);
+    OffsetClock::Instance()->Set(offsetSeconds);
+    zcashdClock = OffsetClock::Instance();
+}
+
+int64_t OffsetClock::GetTime() const {
+    return std::chrono::duration_cast<std::chrono::seconds>(
+            std::chrono::system_clock::now().time_since_epoch()
+            + offsetSeconds
+        ).count();
+}
+
+int64_t OffsetClock::GetTimeMillis() const {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now().time_since_epoch()
+            + offsetSeconds
+        ).count();
+}
+
+int64_t OffsetClock::GetTimeMicros() const {
+    return std::chrono::duration_cast<std::chrono::microseconds>(
+            std::chrono::system_clock::now().time_since_epoch()
+            + offsetSeconds
+        ).count();
+}
+
+int64_t GetTime() {
+    return zcashdClock->GetTime();
+}
+
+int64_t GetTimeMillis() {
+    return zcashdClock->GetTimeMillis();
+}
+
+int64_t GetTimeMicros() {
+    return zcashdClock->GetTimeMicros();
+}
```

Comment:
When a variable is protected by a mutex (as `zcashdClock` is by `cs_clock`), both reads and writes must lock on the mutex. Otherwise the concurrent read is undefined behaviour.

```suggestion
int64_t GetTime() {
    LOCK(cs_clock);
    return zcashdClock->GetTime();
}

int64_t GetTimeMillis() {
    LOCK(cs_clock);
    return zcashdClock->GetTimeMillis();
}

int64_t GetTimeMicros() {
    LOCK(cs_clock);
    return zcashdClock->GetTimeMicros();
}
```

Reply:
I'm a little concerned about the cost of locking here, given that the only things that should ever call any of the functions that swap out the clock reference are in test code. In an ideal world (one where we didn't have to use globals for this sort of thing, but explicitly passed in a clock capability to the operations that needed it) it would be up to the individual clock instance to enforce concurrent access separation to its internal state. If I had my druthers, the mocks wouldn't even be compiled in except for testing.

Reply:
A potential alternative, that would eliminate the lock altogether: what if the clock to use for a given node was specified by a startup parameter, and no other mechanism for setting which clock instance is used on the node is provided? That's closer to the "ideal" behavior I want anyway.

Reply:
Right, but we either need to pay the locking cost or use a different approach.

Compiling out the locking and the mock implementations except when testing *might* work, but note that this means having two versions of zcashd (since the RPC tests use the mocking). That seems undesirable to avoid a very small performance loss.

The other option is to use an atomic.

Reply:
@daira what if the global holding the clock is only set on startup? Would that satisfy your concerns?

Reply:
If it's defined behaviour, yes. In order for it to be defined behaviour there must be no data race (which for example would be true if the shared variable were only set by the init thread).